core/validatorapi/validatorapi.go

```diff
@@ -497,6 +499,103 @@ func (c Component) SubmitBlindedBeaconBlock(ctx context.Context, block *eth2api.
 	return nil
 }
 
+func (c Component) verifyValidatorRegistrationParSig(ctx context.Context, pubKey core.PubKey, slot eth2p0.Slot, randao eth2p0.BLSSignature) error {
+	// Calculate slot epoch
+	epoch, err := c.epochFromSlot(ctx, slot)
+	if err != nil {
+		return err
+	}
+
+	// Randao signing root is the epoch.
+	sigRoot, err := eth2util.EpochHashRoot(epoch)
+	if err != nil {
+		return err
+	}
+
+	return c.verifyParSig(ctx, core.DutyRandao, epoch, pubKey, sigRoot, randao)
+}
+
+func (c Component) verifyValidatorRegistrationSignature(ctx context.Context, registration *eth2api.VersionedSignedValidatorRegistration, pubkey core.PubKey, slot eth2p0.Slot) error {
+
+	epoch, err := c.epochFromSlot(ctx, slot)
+	if err != nil {
+		return err
+	}
+
+	var sig eth2p0.BLSSignature
+	switch registration.Version {
+	case spec.BuilderVersionV1:
+		if registration.V1.Signature == sig {
+			return errors.New("no V1 signature")
+		}
+		sig = registration.V1.Signature
+	default:
+		return errors.New("unknown version")
+	}
+
+	// Verify partial signature
+	// TODO: switch to registration.Root() when implemented on go-eth2-client (PR has been reaised)
+	sigRoot, err := registration.V1.Message.HashTreeRoot()
+	if err != nil {
+		return err
+	}
+
+	return c.verifyParSig(ctx, core.DutyBuilderRegistration, epoch, pubkey, sigRoot, sig)
+}
+
+// SubmitValidatorRegistration receives the partially signed validator (builder) registration.
+func (c Component) SubmitValidatorRegistration(ctx context.Context, registration *eth2api.VersionedSignedValidatorRegistration) error {
+	// Calculate slot epoch
+	// Use timestamp to determine the slot
+	slot := uint64(100000)
+
+	eth2Pubkey := registration.V1.Message.Pubkey
+
+	pubkey, err := core.PubKeyFromBytes(eth2Pubkey[:])
```

Comment:
So this is actually a problem. The pubkey included in the message will be the "pubshare" not the "group/root pubkey" since the VC only knows about it's pubshare not about the group pubkey. 

That means the signatures cannot be aggregated since the pubshare of each VC is different.... 😢 

Add a TODO maybe "this is the pubshare not the group pubkey, oh oh"

Reply:
and btw,`verifyParSig` requires the group pubkey, not the pubshare.